Add Package Registry

KN4CK3R opened the pull request:
This PR adds support for multiple package registries.

Currently implemented are 
- Generic packages
- NuGet
- npm
- Maven
- PyPI

Every repository can be a package registry for all different types of packages. Currently there is no "instance level" registry as not every package type supports it.

Package list:
![list](https://user-images.githubusercontent.com/1666336/126564127-52446673-bf6f-4f81-a95a-0c81e8ac4a98.jpg)

Generic package:
![generic](https://user-images.githubusercontent.com/1666336/126564163-30886a59-4792-4dce-bb57-5fa5c1c0b8b8.jpg)

NuGet:
![nuget](https://user-images.githubusercontent.com/1666336/126564201-5884e1c2-ba19-495f-8f79-290637455a9f.jpg)

npm:
![npm](https://user-images.githubusercontent.com/1666336/126564199-e6366741-5f95-4ba0-b8f8-ad4bd756f910.jpg)

Maven:
![maven](https://user-images.githubusercontent.com/1666336/126564197-70c9fca1-0fee-4511-bdd3-dd567805ff16.jpg)

PyPI:
![pypi](https://user-images.githubusercontent.com/1666336/126564205-59aa91d6-60cf-4c68-965a-44133bddba27.jpg)

close #15706

6543 commented on modules/templates/helper.go:
```diff
@@ -164,7 +165,16 @@ func NewFuncMap() []template.FuncMap {
 		"ReactionToEmoji":                ReactionToEmoji,
 		"RenderNote":                     RenderNote,
-		"IsMultilineCommitMessage":       IsMultilineCommitMessage,
+		"RenderMarkdownToHtml": func(input string) template.HTML {
```
We should have a helper for this

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -1505,6 +1505,40 @@ milestones.filter_sort.most_complete = Most complete
 milestones.filter_sort.most_issues = Most issues
 milestones.filter_sort.least_issues = Least issues
 
+packages = Packages
+packages.desc = Manage repository packages.
+packages.filter.package_type = Type
+packages.filter.package_type.all = All
+packages.published_by = Published %[1]s by <a href="%[2]s">%[3]s</a>
+packages.installation = Installation
+packages.about = About this package
+packages.dependencies = Dependencies
+packages.details = Details
+packages.details.author = Author
+packages.details.project_site = Project Site
+packages.details.license = License
+packages.assets = Assets
+packages.versions = Versions
+packages.versions.on = on
+packages.delete = Delete package
+packages.delete.notice = You are about to delete %s (%s). This operation is irreversible, are you sure?
+packages.delete.success = The package has been deleted.
+packages.generic.use = Download the package, for example with curl:
+packages.generic.documentation = For more information on the generic registry, see <a target="_blank" rel="noopener noreferrer" href="https://docs.gitea.io/en-us/generic">the documentation</a>.
+packages.maven.registry = Setup this registry in your projects <code>pom.xml</code> file:
+packages.maven.use = To use the package include the following in the <code>dependencies</code> block in the <code>pom.xml</code> file:
+packages.maven.use_2 = Alternatively use this Maven command:
+packages.maven.documentation = For more information on the Maven registry, see <a target="_blank" rel="noopener noreferrer" href="https://docs.gitea.io/en-us/maven/">the documentation</a>.
+packages.nuget.registry = Setup this registry with the following command:
+packages.nuget.use = Install the package with the following command:
+packages.nuget.documentation = For more information on the NuGet registry, see <a target="_blank" rel="noopener noreferrer" href="https://docs.gitea.io/en-us/nuget/">the documentation</a>.
+packages.npm.registry = Setup this registry in your projects <code>.npmrc</code> file:
+packages.npm.use = To install the package use the following command:
+packages.npm.documentation = For more information on the npm registry, see <a target="_blank" rel="noopener noreferrer" href="https://docs.gitea.io/en-us/npm/">the documentation</a>.
+packages.pypi.requires = Requires Python
+packages.pypi.use = To install the package use the following command:
+packages.pypi.documentation = For more information on the PyPI registry, see <a target="_blank" rel="noopener noreferrer" href="https://docs.gitea.io/en-us/pypi/">the documentation</a>.
```
Yay! Work for the translators! :laughing:

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -1505,6 +1505,40 @@ milestones.filter_sort.most_complete = Most complete
+packages.npm.use = To install the package use the following command:
+packages.npm.documentation = For more information on the npm registry, see <a target="_blank" rel="noopener noreferrer" href="https://docs.gitea.io/en-us/npm/">the documentation</a>.
+packages.pypi.requires = Requires Python
+packages.pypi.use = To install the package use the following command:
```
I am not sure, but I would say that a comma belongs here.
If that is indeed the case, the same applies to `packages.nuget.use` and `packages.npm.use`.
```ini
packages.pypi.use = To install the package, use the following command:
```
<br>

Additionally, `packages.nuget.use`, `packages.npm.use` and `packages.pypi.use` could potentially all be merged into one string (maybe called `packages.installation.command`) as they all have the exact same meaning (but differ slightly in the wording for NuGet)

delvh reviewed (commented):
Skimmed through it, what I've seen seems pretty good.
My focus was on Maven as that's what I'm most comfortable with.
No doubt, there will be reviewers that actually read through all 5000 lines of code… :wink:

lunny commented:
Since docker is also some kind of package, maybe we should consider what's relation between this one and #14919 . @a1012112796

lunny commented on models/package.go:
```diff
@@ -0,0 +1,338 @@
+func SearchPackages(repositoryID int64, packageType PackageType, query string, skip, take int) (int64, []*Package, error) {
+	cond := builder.NewCond()
+	cond = cond.And(builder.Eq{"repo_id": repositoryID})
+	cond = cond.And(builder.Eq{"type": packageType})
```
```go
var cond builder.Cond = builder.Eq{
    "repo_id": repositoryID,
    "type": packageType,
}
```

lunny commented:
Thanks for the fantastic PR! Two small things you can consider to add in this PR. One is add webhooks for packages, `PackageUploaded`, `PackageUpdated`, and `PackageDeleted`. Another is to add a `package_log` table to record package or files added, updated, deleted.

KN4CK3R commented on modules/packages/maven/metadata.go:
```diff
@@ -0,0 +1,89 @@
+		URL          string `xml:"url"`
+		Distribution string `xml:"distribution"`
+	} `xml:"licenses>license"`
+	Dependencies []struct {
```
I'm no Maven expert but does the client really trust the server informations about the dependencies? I would expect the client downloads the package and checks the local `pom` file for all needed informations.

delvh replied:
As I already said, I am also unsure whether they are needed.
Potentially yes, potentially no.
It could, however, be that the package cannot be registered in case advanced properties are used in the pom file. That's what I'm uncertain about.
In the worst case, this will simply be added to the data model in another PR, after this PR has been merged and tested thoroughly.

KN4CK3R replied:
The `pom` file is not generated but the original file is stored. So all metadata is only used for the Gitea UI. Currently I don't think there is a problem.

KN4CK3R commented:
> Since docker is also some kind of package, maybe we should consider what's relation between this one and #14919 . 

For containers we may follow the OCI spec instead of focusing on Docker alone
https://github.com/opencontainers/distribution-spec

silverwind commented:
<img width="170" alt="image" src="https://user-images.githubusercontent.com/115237/126666236-2745019e-4e3a-46d1-9909-acf9d69b1974.png">

There should be a bit more margin below the headings, please check how it's done in the issue sidebar.

KN4CK3R commented:
> Another is to add a `package_log` table to record package or files added, updated, deleted.

@lunny Why do we need that? How (where) would it look like?

lunny commented:
filter option on search box haven't been selected.

lunny commented on routers/api/v1/api.go:
```diff
@@ -975,6 +981,53 @@ func Routes() *web.Route {
+				m.Group("/packages", func() {
+					m.Group("/generic", func() {
+						m.Group("/{packagename}/{packageversion}/{filename}", func() {
+							m.Get("", generic.DownloadPackageFile)
```
All downloads and read routers needs middleware
`reqRepoReader(models.UnitTypePackages),`

lunny commented:
publish npm package seems return failure.

```
$ more .npmrc
registry=http://localhost:3000/api/v1/repos/lunny/11112/packages/npm/
_authToken=8ddfb644064cd886323xxxxxxxxxxxxxxxxx
```

```
$ npm publish
npm notice
npm notice 📦  npm_publish@1.0.0
npm notice === Tarball Contents ===
npm notice 207B package.json
npm notice === Tarball Details ===
npm notice name:          npm_publish
npm notice version:       1.0.0
npm notice package size:  245 B
npm notice unpacked size: 207 B
npm notice shasum:        4138c2119300283efc504e4c3fee5dbaca2d2cbf
npm notice integrity:     sha512-bBb66Dgn4sPT8[...]xeaDZf2GCfkzQ==
npm notice total files:   1
npm notice
npm ERR! code E400
npm ERR! 400 Bad Request - PUT http://localhost:3000/api/v1/repos/lunny/11112/packages/npm/npm_publish

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/lunny/.npm/_logs/2021-08-15T02_13_03_824Z-debug.log
```

KN4CK3R commented:
> publish npm package seems return failure.

That's a strange behaviour. If `author` is set in your `project.json` the request contains a `"author":{"name":"..."}` object. But if `author` is not set the request contains `"author":""`. In the first case it's an object and in the second a string. Now the json parser is complaining.

KN4CK3R commented:
I added support for RubyGems. Push, install and delete are supported.

![grafik](https://user-images.githubusercontent.com/1666336/130512383-6aa09fd7-daf6-4ede-b545-1ea7348a3522.png)

luwol03 commented:
I tested this feature and found a few things that could be improved:

1. if nothing is found on the search. --> a "no packages matching" message
  ![image](https://user-images.githubusercontent.com/60048565/132127099-43b5019b-a75f-45ed-bf61-7010c66f0712.png)

1. if no package is published and the current logged in user has permissions to publish --> show for each registry a little quick start and the link to the docs like an empty repo shows me how to setup and push commits 
  ![image](https://user-images.githubusercontent.com/60048565/132127122-6a8ca0fc-5d6f-4d6a-9b91-a65fd16795ff.png)
  ![image](https://user-images.githubusercontent.com/60048565/132127161-50feb6cd-5528-4ba4-898b-aea1adc2c7d0.png)

1. a bit more spacing between the insallation instructions 
  ![image](https://user-images.githubusercontent.com/60048565/132127060-64a5131c-cd42-447b-bd04-ba473cf39d76.png) 

1. make the dependencies name links to the dependecies page 
  ![image](https://user-images.githubusercontent.com/60048565/132137821-26144c8a-e54f-47ee-b903-4bd006db11e0.png)

1. add a download counter like npmjs.com have it.
  ![image](https://user-images.githubusercontent.com/60048565/132137876-372888e8-227e-4d92-8a1b-ec9f674993a7.png)

1. maybe a size and total files counter would also be an option.
  ![image](https://user-images.githubusercontent.com/60048565/132137912-c72fed0a-edb7-43bc-91a3-8c690f0482d7.png)

1. why does the generic package dont show the README.md?

1. `devDependencies` and `peerDependencies` defined in the `package.json` dont show in the Dependencies section ![image](https://user-images.githubusercontent.com/60048565/132138629-c2669c86-6c4a-4bf7-a09b-59c79f87fb71.png)
  ```json
  {
    ...
    "dependencies": {
      "@material-ui/core": "^4.12.3",
      "express": "^4.17.1"
    },
    "peerDependencies": {
      "jsonwebtoken": "8.5.1"
    },
    "optionalDependencies": {
      "hello": "0.3.2"
    },
    "devDependencies": {
      "koa": "^2.13.1"
    }
  }
  ```

9. show also the package.jsons keywords in the sidebar

KN4CK3R commented:
> if nothing is found on the search. --> a "no packages matching" message

There is now a "nothing found" message

> if no package is published

I added a link to the package docs.

> a bit more spacing between the insallation instructions

done

> make the dependencies name links to the dependecies page

I don't think that's possible because the dependencies may not be hosted by Gitea

> add a download counter like npmjs.com have it.

I don't know if we want that feature. And if we do, how fine grained it should be. For example a simple download counter or "today", "week", ... As there are package types with multiple files I don't know at the moment how to count the downloads. If every file download adds one, for two files the package download counter would increase by two.

> maybe a size and total files counter would also be an option.

done

> why does the generic package dont show the README.md and dont allow to upload multiple assets?

Readme support is not planned. Multiple assets may be allowed. This may be a problem for a potential download counter.

> `devDependencies` and `peerDependencies` defined in the `package.json` dont show in the Dependencies section

done

> show also the package.jsons keywords in the sidebar

done for npm

> allow to download packages without authentication

done

luwol03 commented:
I want to say again thank you very much for that great feature.

> > make the dependencies name links to the dependecies page
> 
> I don't think that's possible because the dependencies may not be hosted by Gitea

Yeah, you’re right. Firstly I thought about links to npmjs.com, but as you said not all packages are hosted there and the user can also have changed his registry settings.

> > add a download counter like npmjs.com have it.
> 
> I don't know if we want that feature. And if we do, how fine grained it should be. For example a simple download counter or "today", "week", ... As there are package types with multiple files I don't know at the moment how to count the downloads. If every file download adds one, for two files the package download counter would increase by two.

I think all package managers only have one endpoint which can be used to count a package as a Download + 1. For the generic package you could use the sum of all files. And for the visualization you could also take a look at githubs kind of chart. I think this also looks pretty good.

> > why does the generic package dont show the README.md and dont allow to upload multiple assets?
> 
> Readme support is not planned. Multiple assets may be allowed. This may be a problem for a potential download counter.

Why is that not the planed? A npm package also shows me a readme.

> > show also the package.jsons keywords in the sidebar
> 
> done for npm

Does such a information also exists on other package managers?

> > allow to download packages without authentication
> 
> done

Is this for all packages, or is there an option for the repo owner?

I now have two questions left:
1. Could you may update the Screenshots? 
2. What is with docker packages?

redstonedesigner commented:
This seems like a great feature, and something I cannot wait to see implemented.

One thing, however - Does this handle the generation of maven-metadata.xml files?  I use Renovate on my Gitea instance and I currently have a Sonatype Nexus 3 OSS deployed with the Maven repository.  Renovate refused to work until the maven-metadata.xml files were generated.

KN4CK3R commented:
> > > add a download counter like npmjs.com have it.
> > 
> > 
> > I don't know if we want that feature. And if we do, how fine grained it should be. For example a simple download counter or "today", "week", ... As there are package types with multiple files I don't know at the moment how to count the downloads. If every file download adds one, for two files the package download counter would increase by two.
> 
> I think all package managers only have one endpoint which can be used to count a package as a Download + 1. For the generic package you could use the sum of all files. And for the visualization you could also take a look at githubs kind of chart. I think this also looks pretty good.

Yes, there is only one endpoint but it gets called multiple times for a single package download. For example Maven: 2-3 hits (x.jar, x.pom, maven-metadata.xml), Python (x.whl, x.tar.gz). In these cases a simple +1 does not work.

> > > why does the generic package dont show the README.md and dont allow to upload multiple assets?
> > 
> > 
> > Readme support is not planned. Multiple assets may be allowed. This may be a problem for a potential download counter.
> 
> Why is that not the planed? A npm package also shows me a readme.

Most packages contain metadata. The npm package manager embeds the whole readme file into the metadata. This is displayed on the detail page, not the readme file contained in the package itself.

> > > show also the package.jsons keywords in the sidebar
> > 
> > 
> > done for npm
> 
> Does such a information also exists on other package managers?

Nuget can have `tags` in the package metadata.

> > > allow to download packages without authentication
> > 
> > 
> > done
> 
> Is this for all packages, or is there an option for the repo owner?

For all packages. If anonymous access it not allowed for a repository it is not allowed for packages too.

> 2. What is with docker packages?

I have a (unpublished) PR which adds support for containers. It passes all OCI conformance tests and therefore works with every OCI client, docker included. But because the data structure does not match with the "normal" packages I will not add it into this PR. It's large enough 😉

KN4CK3R commented:
> One thing, however - Does this handle the generation of maven-metadata.xml files? I use Renovate on my Gitea instance and I currently have a Sonatype Nexus 3 OSS deployed with the Maven repository. Renovate refused to work until the maven-metadata.xml files were generated.

The "what versions are available" `maven-metadata.xml` gets generated on the fly. The snapshot `maven-metadata.xml` is just the provided one by maven when publishing a package.
You could test it and tell me if it works or not.

luwol03 commented:
> Yes, there is only one endpoint but it gets called multiple times for a single package download. For example Maven: 2-3 hits (x.jar, x.pom, maven-metadata.xml), Python (x.whl, x.tar.gz). In these cases a simple +1 does not work.

But there should be a mechanism to count that. Gh and npm already have such statistics.
![image](https://user-images.githubusercontent.com/60048565/133147219-2432bf04-dbe2-4819-b9a9-74066bfe4f94.png)
![image](https://user-images.githubusercontent.com/60048565/133147308-33c0b6c5-71c8-4709-b2d4-44fbafd053f3.png)

> For all packages. If anonymous access it not allowed for a repository it is not allowed for packages too.

Maybe there should also be a option to only make the packages/wiki/issues public?

> I have a (unpublished) PR which adds support for containers. It passes all OCI conformance tests and therefore works with every OCI client, docker included. But because the data structure does not match with the "normal" packages I will not add it into this PR. It's large enough :wink:

Good to know. I am already looking forward to it.

lunny commented on models/package.go:
```diff
@@ -0,0 +1,382 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
```
How about move this file into `models/package/package.go` so that more package related models could be added easier.

KN4CK3R replied:
done

gp-Airee commented:
Sounds like scope creep to me. I appreciate the narrow scope and purpose of Gitea and wouldn't want this merged. If someone wants a package registry, there are standalone solutions that can be implemented.

markkrj commented:
> Sounds like scope creep to me. I appreciate the narrow scope and purpose of Gitea and wouldn't want this merged. If someone wants a package registry, there are standalone solutions that can be implemented.

Providing built source code is not that far outside the scope of a source code management system...

gp-Airee commented:
You said it yourself, it's a source code management system. This isn't source code. Gitea shouldn't be managing it. There's not even a built in CI system for Gitea to build the packages.

If there were a precedent set for Gitea including devops features, like a CI system, this would make sense. But it's not making sense to me with what Gitea is currently.

-------- Original Message --------
On Sep 28, 2021, 9:16 PM, Marcos de Oliveira wrote:

>> Sounds like scope creep to me. I appreciate the narrow scope and purpose of Gitea and wouldn't want this merged. If someone wants a package registry, there are standalone solutions that can be implemented.
>
> Providing built source code is not that far outside the scope of a source code management system...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, [view it on GitHub](https://github.com/go-gitea/gitea/pull/16510#issuecomment-929766383), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AD67NIYYC3B577PQDPZVXNLUEJZIJANCNFSM5AYZVHDQ).

lunny commented:
> You said it yourself, it's a source code management system. This isn't source code. Gitea shouldn't be managing it. There's not even a built in CI system for Gitea to build the packages. If there were a precedent set for Gitea including devops features, like a CI system, this would make sense. But it's not making sense to me with what Gitea is currently.
> […](#)
> -------- Original Message --------
> On Sep 28, 2021, 9:16 PM, Marcos de Oliveira wrote: > Sounds like scope creep to me. I appreciate the narrow scope and purpose of Gitea and wouldn't want this merged. If someone wants a package registry, there are standalone solutions that can be implemented. Providing built source code is not that far outside the scope of a source code management system... — You are receiving this because you commented. Reply to this email directly, [view it on GitHub]([#16510 (comment)](https://github.com/go-gitea/gitea/pull/16510#issuecomment-929766383)), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AD67NIYYC3B577PQDPZVXNLUEJZIJANCNFSM5AYZVHDQ).

Or we can have a configuration to disable the feature.

gp-Airee commented:
> > You said it yourself, it's a source code management system. This isn't source code. Gitea shouldn't be managing it. There's not even a built in CI system for Gitea to build the packages. If there were a precedent set for Gitea including devops features, like a CI system, this would make sense. But it's not making sense to me with what Gitea is currently.
> > […](#)
> > -------- Original Message --------
> > On Sep 28, 2021, 9:16 PM, Marcos de Oliveira wrote: > Sounds like scope creep to me. I appreciate the narrow scope and purpose of Gitea and wouldn't want this merged. If someone wants a package registry, there are standalone solutions that can be implemented. Providing built source code is not that far outside the scope of a source code management system... — You are receiving this because you commented. Reply to this email directly, [view it on GitHub]([#16510 (comment)](https://github.com/go-gitea/gitea/pull/16510#issuecomment-929766383)), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AD67NIYYC3B577PQDPZVXNLUEJZIJANCNFSM5AYZVHDQ).
> 
> Or we can have a configuration to disable the feature.

That's really the bare minimum you could do if you ask me. Gitea's greatest strength is it's simplicity.

I urge the maintainers of this project to look closely at the features implemented to date and decide whether this is in line with any feature implemented into Gitea to date.

If you come to the same conclusion that I did, that this is a strong departure from the purpose of Gitea, I personally hope that this is separated from the regular Gitea releases, or that an extension system is developed and this is implemented through it.

markkrj commented:
Look, people love the simplicity of gitea, and that's exactly why features like this are so much loved. It simplify a lot the life of developers. Package registries are good for distributing end-user apps, but are specially important for providing libraries for such apps... Having both the code and needed libraries in the same place, with the same authentication and APIs... That's game changer! Not just for developers but for sysadmins also, as we can replace 2-3 another softwares by gitea alone. No need to setup another system with a different database, authentication and policies, integrate it and always take extra care not to break such integration. That's simplicity to me.

gp-Airee commented:
That's not simplicity to me.

> package registries are good for distributing end-user apps
If you say so, but it's definitely the wrong tool for the job. Uploading it to a CDN via your CI pipeline is a better approach. Again, both things that Gitea has made no attempt to implement.

> and needed libraries
Needed libraries for what? Building the code? Something Gitea doesn't do?

lunny commented:
> > package registries are good for distributing end-user apps
> 
> If you say so, but it's definitely the wrong tool for the job. Uploading it to a CDN via your CI pipeline is a better approach. Again, both things that Gitea has made no attempt to implement.
> 
> > and needed libraries
> 
> Needed libraries for what? Building the code? Something Gitea doesn't do?
> 
> That's not simplicity to me. This is taking an extremely focused project to serve a specific purpose and pushing it in the direction of a monolithic suite of tools for the greater development workflow.

If you are in a company and you want to store your private npm/python/maven packages in somewhere, and reuse them in your code. (Currently, gitea supports golang and brew because they only depends on git and a slight change). Then you have to install all kinds of tools. This PR makes an all in one package registry. And you will feel no pain for all these things.
Of course, if you have this requirement to publish your private packages. You could disable it(when this PR changed), it should not consume any CPU or disk like there seems be no this feature of Gitea.

KN4CK3R commented:
I talked with @lunny the last days about the package registry "scope". Currently every repository has its own independent registry. The alternative would be a user/org registry. I have listed some pro/cons we should consider to make a decision.

Note: GitHub supports only user/org registries (with packages associated to a repository) while GitLab uses repository registries. GitLab allows to use an user/org endpoint for some package types.

### GitHub
For GitHub every package (except containers) must be associated with a repository. Because only an user/org endpoint is available, the reference for a repository must be set in the package metadata (npm: `"repository": "https://github.com/user/repo"`). If you try to push a second image with the same name but a reference to another repository, you get an error message like `Package xxx is already associated with another repository`. 

### GitLab
For GitLab every package is associated with a repository because the repository is the registry. Because of this you do not have to specify a repository reference in the package metadata. This allows to push third-party packages to the registry where you can't control the metadata. You can push the same package to multiple repository-registries without problems because they are independent. This leads to problems with user/org endpoints because there are now multiple packages with the same name (and maybe different versions).

Note 2: The need for the repository-association is the inherited permission model of the repository. If a user has access to the repository, he has access to the packages. For GitHub container-packages not associated with a repository, you can choose if the visibility is public or private.

| | GitHub | GitLab |
| - | - | - |
| Pro | - Every package is available from one endpoint<br>- No name collissions | - Repository reference not needed |
| Contra | - Repository reference in package metadata needed | - Name collissions with user/org endpoints<br>- Duplicated endpoints (user/org scope, registry scope) |

Currently implemented is the GitLab version without the user/org endpoints, so there are no name collissions. The downside is you may have to specify multiple package registries in a project because the dependency packages may be spread over multiple repositories (the alternative would be a single repository which contains all packages).

For Gitea we may not enforce the association to a repository when publishing a package to the user/org endpoint. In this case we could allow to manually set a repository reference later in the UI (GitHub allows this for unassociated containers). Because a repository reference is missing, there is no permission model. Maybe all these packages should be private for the user/org then. This would get us all pros and may prevent to contras. An edge case would be a manually associated package for which you publish a new version without repository reference. Would that package be automatically associated with the repository of the present package version?

wxiaoguang commented:
I think the GitLab model is better. The Pro of GitHub model (Con of GitLab model) can not be a problem:

1. The endpoint problem: users can setup a separate repository for package purpose.
2. The name collision problem: as 1, a special package repository also works.

KN4CK3R commented:
In https://github.com/KN4CK3R/gitea/tree/feature-packages-owner I have implemented the Github model with the user/org package registry.
I did not add the ui yet because I'm not sure about the urls to use.
The packages of a user may be visible at `https://gitea.io/USERNAME?tab=packages` but where should the package view/settings/... be located? `gitea.io/USERNAME/packages/...` conflicts with a potential repository `packages`. Github uses `github.com/users/USERNAME/packages` and `github.com/orgs/ORGNAME/packages`. The Gitlab way may something like `gitea.io/USERNAME/-/packages/`. `-` must be a reserved repository name then but allows every extension in the future. What do you think?

lunny commented:
> In https://github.com/KN4CK3R/gitea/tree/feature-packages-owner I have implemented the Github model with the user/org package registry. I did not add the ui yet because I'm not sure about the urls to use. The packages of a user may be visible at `https://gitea.io/USERNAME?tab=packages` but where should the package view/settings/... be located? `gitea.io/USERNAME/packages/...` conflicts with a potential repository `packages`. Github uses `github.com/users/USERNAME/packages` and `github.com/orgs/ORGNAME/packages`. The Gitlab way may something like `gitea.io/USERNAME/-/packages/`. `-` must be a reserved repository name then but allows every extension in the future. What do you think?

I like GitLab's, looks like it supports more features.

KN4CK3R commented:
4162676 is the last commit with the old repository level package registry.

Now all packages are at user/org level and can be linked to a repository in the settings:
![package_settings](https://user-images.githubusercontent.com/1666336/141615603-7148aebb-69e0-42ce-8b4d-bfdc51b066b9.jpg)

If linked to a repository there are some additional options in the ui:
![package_link](https://user-images.githubusercontent.com/1666336/141615621-8016d7c2-5430-4b7a-b049-b3589a7833c5.jpg)

Also there is a download counter now.

lunny commented on docs/content/doc/packages/npm.en-us.md:
````diff
@@ -0,0 +1,108 @@
+To register the project’s package registry you need to configure a new package source.
+
+```shell
+npm config set {scope}:registry https://gitea.example.com/api/v1/packages/{owner}/npm/
````
```npm config set {scope}:registry https://gitea.example.com/{owner}```

how about to reduce the package url length?

KN4CK3R commented:
To help testing I created a list with steps how to create and publish a minimal package of every type.

{username} = your username
{password} = your password
{token} = your personal access token

The examples use the default `http://localhost:3000/` url.

## Composer

```shell
composer init

zip -r php-package.zip .

curl --user {username}:{password} --upload-file php-package.zip http://localhost:3000/api/v1/packages/{username}/composer?version=1.0.3
```

## Generic

```shell
curl --user {username}:{password} --upload-file file.bin http://localhost:3000/api/v1/packages/{username}/generic/generic-package/1.0.3/file.bin

curl --user {username}:{password} http://localhost:3000/api/v1/packages/{username}/generic/generic-package/1.0.3/file.bin
```

## Maven

```shell
mvn archetype:generate -DgroupId=io.gitea -DartifactId=maven-package -DarchetypeArtifactId=maven-archetype-quickstart

nano ~/.m2/settings.xml
```
```xml
<settings>
  <servers>
    <server>
      <id>gitea</id>
      <configuration>
        <httpHeaders>
          <property>
            <name>Authorization</name>
            <value>token {token}</value>
          </property>
        </httpHeaders>
      </configuration>
    </server>
  </servers>
</settings>
```
```shell
cd maven-package

nano pom.xml
```
```xml
<project ...>
  ...
  <repositories>
    <repository>
      <id>gitea</id>
      <url>http://localhost:3000/api/v1/packages/{owner}/maven</url>
    </repository>
  </repositories>
  <distributionManagement>
    <repository>
      <id>gitea</id>
      <url>http://localhost:3000/api/v1/packages/{owner}/maven</url>
    </repository>
    <snapshotRepository>
      <id>gitea</id>
      <url>http://localhost:3000/api/v1/packages/{owner}/maven</url>
  </snapshotRepository>
  </distributionManagement>
</project>
```
```shell
mvn deploy
```

You may encounter errors like `Source option 5 is no longer supported. Use 6 or later.` or `Target option 1.5 is no longer supported. Use 1.6 or later.`. This can be fixed with an addition to the `pom.xml`:
```xml
<properties>
  <maven.compiler.source>1.6</maven.compiler.source>
  <maven.compiler.target>1.6</maven.compiler.target>
</properties>
```

## npm

If you use a package `{scope}` (example: `@test`), you need to use `@test/package-name` as package name in `npm init`.

```shell
npm config set {scope}:registry http://localhost:3000/api/v1/packages/{username}/npm/

npm config set -- '//localhost:3000/api/v1/packages/{username}/npm/:_authToken' "{token}"

npm init

npm publish
```

## NuGet

```shell
dotnet nuget add source --name gitea --username {username} --password {password} http://localhost:3000/api/v1/packages/{username}/nuget/index.json

mkdir nuget-package

dotnet new classlib

nano nuget-package.csproj
```
```xml
<PropertyGroup>
  <PackageId>nuget-package</PackageId>
  <Version>1.0.3</Version>
  <Authors>...</Authors>
  <Company>...</Company>
  ...
</PropertyGroup>
```
```shell
dotnet pack

dotnet nuget push --source gitea .\bin\Debug\nuget-package.1.0.3.nupkg
```

## PyPI

```shell
nano ~/.pypirc
```
```ini
[distutils]
index-servers = gitea

[gitea]
repository = http://localhost:3000/api/v1/packages/{username}/pypi
username = {username}
password = {password}
```
```shell
mkdir pypi-package

nano pypi-package/__init__.py
```
```
def Dummy():
	print("Gitea Package")
	return
```
```shell
nano pypi-package/setup.py
```
```python
import setuptools

setuptools.setup(
    name="pypi-package",
    version="1.0.3",
    author="Gitea Author",
    author_email="author@gitea.io",
    description="Gitea Package",
    packages=setuptools.find_packages(),
    classifiers=[
        "Programming Language :: Python :: 3",
        "License :: OSI Approved :: MIT License",
        "Operating System :: OS Independent",
    ],
    python_requires='>=3.6',
)
```
```shell
python3 -m twine upload --repository gitea dist/*
```

## RubyGems

```shell
nano ~/.gem/credentials
```
```
---
http://localhost:3000/api/v1/packages/{username}/rubygems: Bearer {token}
```
```shell
nano ruby-package.gemspec
```
```
Gem::Specification.new do |s|
  s.name        = 'ruby-package'
  s.version     = '1.0.3'
  s.summary     = "Gitea Package"
  s.description = "Gitea Package"
  s.authors     = ["Gitea"]
  s.email       = 'package@gitea.to'
  s.files       = ["lib/main.rb"]
  s.homepage    = 'https://gitea.io'
  s.license     = 'MIT'
end
```
```shell
mkdir lib

nano lib/main.rb
```
```
class GiteaPackage
  def self.dummy
    puts "Gitea!"
  end
end
```
```shell
gem build ruby-package.gemspec

gem push --host http://localhost:3000/api/v1/packages/KN4CK3R/rubygems ruby-package-1.0.3.gem
```

lunny commented:
When publish the same npm package to two different scope on different owner, the error occurs.

```
2021/11/24 17:06:31 ...packages/packages.go:187:addFileToPackageVersion() [T] Adding package file: 3, npmpublish-1.1.0.tgz
2021/11/24 17:06:31 ...ages/package_blob.go:37:GetOrInsertBlob() [I] [SQL] SELECT "id", "size", "hash_md5", "hash_sha1", "hash_sha256", "hash_sha512" FROM "package_blob" WHERE "size"=$1 AND "hash_md5"=$2 AND "hash_sha1"=$3 AND "hash_sha256"=$4 AND "hash_sha512"=$5 LIMIT 1 [334 0de09f2288944be373a8ca57a682fce6 53c4c4ca02f36a3054569d7137c95a59f025d5c7 d4974924d8e56a351049d3d213e1e96315b0117d71331e28c2ac57d2024f827e 46ad4ef8be8ca3dfcf63f47149783ab08e2d54b23bf6548b86c84d7d0a078314a2930e00035ae19a2435e142d2989a36a39666a0f4a0d93290ab2819e1aa8645] - 2.472085ms
2021/11/24 17:06:31 ...packages/packages.go:197:addFileToPackageVersion() [I] [SQL] INSERT INTO "package_blob" ("size","hash_md5","hash_sha1","hash_sha256","hash_sha512") VALUES ($1,$2,$3,$4,$5) RETURNING "id" [334 0de09f2288944be373a8ca57a682fce6 53c4c4ca02f36a3054569d7137c95a59f025d5c7 d4974924d8e56a351049d3d213e1e96315b0117d71331e28c2ac57d2024f827e 46ad4ef8be8ca3dfcf63f47149783ab08e2d54b23bf6548b86c84d7d0a078314a2930e00035ae19a2435e142d2989a36a39666a0f4a0d93290ab2819e1aa8645] - 1.765292ms
2021/11/24 17:06:31 ...packages/packages.go:205:addFileToPackageVersion() [E] Error saving package blob in content store: Object name already exists as a directory.
2021/11/24 17:06:31 .../packages/npm/npm.go:131:UploadPackage() [I] [SQL] ROLLBACK [] - 1.799636ms
2021/11/24 17:06:31 .../packages/npm/npm.go:34:apiError() [E] Object name already exists as a directory.
2021/11/24 17:06:31 Completed PUT /api/v1/packages/lunny3/npm/@4444%2fnpmpublish 500 Internal Server Error
```

KN4CK3R commented:
I have added an admin page to manage packages.

![grafik](https://user-images.githubusercontent.com/1666336/144319911-faefbb40-3034-4c60-b79d-279811240439.png)

qwerty287 commented on templates/package/content/maven.tmpl:
```diff
@@ -0,0 +1,71 @@
+{{if eq .PackageDescriptor.Package.Type "maven"}}
+	<h4 class="ui top attached header">{{.i18n.Tr "packages.installation"}}</h4>
```
Maven repositories are not only usable with Maven itself, but also with Gradle, sbt, Leiningen... It would be nive if there could be a similar screen to the one of [Jitpack](https://jitpack.io) that makes it easier to copy-paste the repos and deps to the build files.

View on Jitpack:
![Screenshot 2021-12-06 at 16-12-19 JitPack Publish JVM and Android libraries](https://user-images.githubusercontent.com/80460567/144871594-b4fe7d7b-2e19-4423-b78a-c0b46ed8c746.png)

KN4CK3R commented:
Added [Conan](https://conan.io) package registry for C++ packages. Conan was a little beast because of the revision system which allows multiple sub-versions per version. My implementation has full support for v2 revisions.

![conan](https://user-images.githubusercontent.com/1666336/153951243-eb65ff49-950c-4601-99d4-d4586d52e5c0.png)

lunny commented on docs/content/doc/features/comparison.en-us.md:
```diff
@@ -48,7 +48,8 @@ _Symbols used in table:_
 | Repository Tokens with write rights | ✓                                                  | ✘    | ✓         | ✓         | ✓         | ✓              | ✓            |
-| Built-in Container Registry         | [✘](https://github.com/go-gitea/gitea/issues/2316) | ✘    | ✘         | ✓         | ✓         | ✘              | ✘            |
+| Built-in Container Registry         | [✘](https://github.com/go-gitea/gitea/issues/2316) | ✘    | ✓         | ✓         | ✓         | ✘              | ✘            |
+| Built-in Package Registry           | ✓                                                  | ✘    | ✓         | ✓         | ✓         | ✘              | ✘            |
```
Could we have more detail abouth which packages Gitea supports? And Container should be one of them?

KN4CK3R replied:
At the moment there is no container registry but yes, should be one entry.
Which package is supported is documented in [doc/packages/overview.en-us.md](https://github.com/go-gitea/gitea/pull/16510/files#diff-77c07ab67c30445268eef68a3876876504fc18da985ff3b7fb6a065f9e814c04)

lunny commented on models/migrations/v210.go:
```diff
@@ -0,0 +1,81 @@
+		ID               int64  `xorm:"pk autoincr"`
+		OwnerID          int64  `xorm:"UNIQUE(s) INDEX NOT NULL"`
+		RepoID           int64  `xorm:"INDEX"`
+		Type             string `xorm:"UNIQUE(s) INDEX NOT NULL"`
```
type could be a limited size, i.e. 16 or 32

lunny commented on models/migrations/v210.go:
```diff
@@ -0,0 +1,81 @@
+		PackageID     int64 `xorm:"UNIQUE(s) INDEX NOT NULL"`
+		CreatorID     int64
+		Version       string
+		LowerVersion  string             `xorm:"UNIQUE(s) INDEX NOT NULL"`
```
Should version have a limit size?

silverwind commented:
Tagging for review later, I guess I can at least test the npm parts.

KN4CK3R commented:
To keep it all together I added the container registry too. It follows the OCI spec and is (currently) usable with Docker and Helm Charts. All other OCI compliant images should be usable too but there may be no metadata extraction.

<details>
  <summary>Screenshots</summary>

Docker Image:
![grafik](https://user-images.githubusercontent.com/1666336/159169391-d523583e-6010-48da-9b42-e23674ef7f60.png)

MultiArch Docker Image:
![grafik](https://user-images.githubusercontent.com/1666336/159169400-42cf5936-039b-4008-915f-f730bb62e72b.png)

Image Tag List:
![grafik](https://user-images.githubusercontent.com/1666336/159169420-81296602-5b5e-43f7-84a8-9e9fcc4d33ee.png)
![grafik](https://user-images.githubusercontent.com/1666336/159169439-0116f4a4-ee4a-4c45-8710-0ef0f5d631b6.png)

Helm Chart:
![grafik](https://user-images.githubusercontent.com/1666336/159169457-86364ca4-8df4-464a-a2a6-d6465a8effaa.png)
</details>

KN4CK3R commented:
I have updated the instructions in https://github.com/go-gitea/gitea/pull/16510#issuecomment-976344348 for the new url route and added Conan and Container (Docker). Please test this PR and send me feedback.

lunny commented:
```
PANIC: template: package/content/npm:28: function "RenderMarkdownToHtml" not defined
```

wxiaoguang commented:
Since the `ContextUser` PR has been merged, will this PR use the context user?

* https://github.com/go-gitea/gitea/pull/18798

KN4CK3R commented:
Yes, I will merge in a moment.

wxiaoguang commented on modules/context/context.go:
```diff
@@ -359,31 +365,41 @@ func (ctx *Context) ServeFile(file string, names ...string) {
 
+// UploadStream returns the request body or the first form file
+// Only form files need to get closed.
+func (ctx *Context) UploadStream() (io.ReadCloser, bool, error) {
```
```suggestion
func (ctx *Context) UploadStream() (rd io.ReadCloser, needClose bool, err error) {
```

wxiaoguang commented on modules/context/package.go:
```diff
@@ -0,0 +1,109 @@
+func PackageContexter() func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+			ctx := Context{
+				Resp: NewResponse(resp),
+				Data: map[string]interface{}{},
+			}
+
+			ctx.Req = WithContext(req, &ctx)
```
I think here we need a new PackageContext and make `web/route.go` to support it .... otherwise the `Context` is too heavy, it has too many unrelated fields.

KN4CK3R replied:
First I wanted to add a new `PackageContext` but I ran into too many problems because lots of methods expect the current context. I would have to add new middlewares, change the router and so on. At the end I thought it is not worth it.

wxiaoguang replied:
I see ...... then maybe the `Context` system also need to be refactored later. 😂

wxiaoguang commented on modules/context/repo.go:
```diff
@@ -441,6 +441,7 @@ func RepoAssignment(ctx *Context) (cancel context.CancelFunc) {
 	ctx.Repo.Owner = owner
 	ctx.ContextUser = owner
 	ctx.Data["Username"] = ctx.Repo.Owner.Name
+	ctx.ContextUser = owner
```
```suggestion
	ctx.ContextUser = owner
	ctx.Data["Username"] = ctx.Repo.Owner.Name
```

Duplicated `	ctx.ContextUser = owner`

wxiaoguang reviewed (commented):
Partially review, some small nits. (to be continued ....)

wxiaoguang commented on routers/api/packages/api.go:
```diff
@@ -0,0 +1,397 @@
+		)
+
+		// Manual mapping of routes because {image} can contain slashes which chi does not support
+		r.Route("/*", "HEAD,GET,POST,PUT,PATCH,DELETE", func(ctx *context.Context) {
```
It had better to be a named function, instead of an anonymous function

KN4CK3R replied:
But it's only purpose is the be called in this router. I don't think it adds much benefit to give it the name `slashAwareContainerRouter`.

wxiaoguang replied:
But it could keep the file `routers/api/packages/api.go` clear and simple. I think we should not mix the handler code and router definition together.

KN4CK3R replied:
That's not handler code but a router definition? The manual way because chi can't handle `/` in a parameter.

wxiaoguang commented on routers/web/user/package.go:
```diff
@@ -0,0 +1,341 @@
+		ctx.ServerError("GetPackageByName", err)
+		return
+	}
+	if len(pvs) == 0 {
+		ctx.NotFound("", err)
+	}
```
`return` missed. (here and below, maybe more)

We really need a framework support to do `return ctx.RespondServerError(...)` to avoid such low-level mistakes.

wxiaoguang commented on templates/package/content/maven.tmpl:
```diff
@@ -0,0 +1,71 @@
+		<div class="ui form">
+			<div class="field">
+				<label>{{svg "octicon-code"}} {{.i18n.Tr "packages.maven.registry" | Safe}}</label>
+				<div class="markup"><pre class="code-block"><code>&lt;repositories&gt;
+	&lt;repository&gt;
+		&lt;id&gt;gitea&lt;/id&gt;
+		&lt;url&gt;{{AppUrl}}api/packages/{{.PackageDescriptor.Owner.Name}}/maven&lt;/url&gt;
```
We can use the `template` trick (with JS) to make code more readable. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template

KN4CK3R replied:
I would agree if this text would change frequently. As it is static I would prefer to make it usable without JS.

KN4CK3R commented on models/packages/descriptor.go:
```diff
@@ -0,0 +1,188 @@
+		metadata = &maven.Metadata{}
+	case TypePyPI:
+		metadata = &pypi.Metadata{}
+	case TypeRubyGems:
+		metadata = &rubygems.Metadata{}
+	}
+	if metadata != nil {
+		if err := json.Unmarshal([]byte(pv.MetadataJSON), &metadata); err != nil {
+			return nil, err
+		}
+	}
```
That if is the "default" case for the switch. If everything runs well (no manual database change or something), metadata may never be nil here because every implemented package type is in the list.

wxiaoguang replied:
Personally I think it's worthy to make the switch-case can cover all edge-cases, at least add an error log or panic.

In future if a new package type is added, an error (or panic) here would help developers to find the potential bug if the new `case` branch is forgotten. Most golang standard packages also does panic if there is an unexpected case.

wxiaoguang replied:
It seems that the CI catches an unhandled case: `unknown package type: generic`  😂

KN4CK3R replied:
The "generic" case was probably the reason for the `if metadata != nil` check. I have added that case now.

KN4CK3R commented on models/packages/container/search.go:
```diff
@@ -0,0 +1,222 @@
+
+// GetImageTags gets a sorted list of the tags of an image
+// The result is suitable for the api call.
+func GetImageTags(ctx context.Context, ownerID int64, image string, n *int, last string) ([]string, error) {
```
`n` and `last` are from the OCI spec: https://github.com/opencontainers/distribution-spec/blob/main/spec.md#content-discovery

> When n is zero, this endpoint MUST return an empty list

wxiaoguang replied:
Does the OCI spec define `-1`? The `n` is only processed by one of our routers, `-1` can be handled easily.

KN4CK3R commented on models/migrations/v212.go:
```diff
@@ -0,0 +1,94 @@
+	type PackageBlobUpload struct {
+		ID             string             `xorm:"pk"`
+		BytesReceived  int64              `xorm:"NOT NULL DEFAULT 0"`
+		HashStateBytes []byte             `xorm:"BLOB"`
```
So, should I change it or not? `Session` uses `BLOB` too.

wxiaoguang replied:
I prefer not to use blob, it's difficult to be maintained in database.

Just my opinion, not a blocker. Either is fine since it works.

wxiaoguang commented on routers/api/packages/npm/npm.go:
```diff
@@ -0,0 +1,309 @@
+
+// packageNameFromParams gets the package name from the url parameters
+// Variations: /name/, /@scope/name/, /@scope%2Fname/
+func packageNameFromParams(ctx *context.Context) (string, error) {
+	scope := ctx.Params("scope")
+	id := ctx.Params("id")
+	if scope != "" {
+		return fmt.Sprintf("@%s/%s", scope, id), nil
+	}
+	return url.QueryUnescape(id)
```
`QueryUnescape` seems incorrect. IIRC `Params` should already do the PathUnescape.

And the `fmt.Sprintf("@%s/%s", scope, id)` above doesn't do any escaping

KN4CK3R replied:
> `QueryUnescape` seems incorrect. IIRC `Params` should already do the PathUnescape.

Tested it again and it's not needed.

> And the `fmt.Sprintf("@%s/%s", scope, id)` above doesn't do any escaping

And it should not. This method gets called from multiple endpoints: `/{id}/` and `/@{scope}/{id}/`
If it's the first route the name is only the `id`, if it's the second route it's `@scope/id`.

wxiaoguang commented on routers/api/packages/npm/api.go:
```diff
@@ -0,0 +1,73 @@
+		Dist: npm_module.PackageDistribution{
+			Shasum:    pd.Files[0].Blob.HashSHA1,
+			Integrity: "sha512-" + base64.StdEncoding.EncodeToString(hashBytes),
+			Tarball:   fmt.Sprintf("%s/%s/-/%s/%s", registryURL, url.QueryEscape(pd.Package.Name), pd.Version.Version, pd.Files[0].File.LowerName),
```
The PackageName and FileLowerName should be PathEscape-ed

wxiaoguang commented on routers/api/packages/generic/generic.go:
```diff
@@ -0,0 +1,164 @@
+		}
+		apiError(ctx, http.StatusInternalServerError, err)
+		return
+	}
+}
```
`ctx.Status` lost?

KN4CK3R replied:
Not needed but I added a `StatusOK`.

wxiaoguang commented on modules/util/filebuffer/file_backed_buffer.go:
```diff
@@ -0,0 +1,147 @@
+		n, err = b.file.Write(p)
+	} else {
+		if b.size+int64(len(p)) > b.maxMemorySize {
+			b.file, err = os.CreateTemp("", "buffer-")
```
```suggestion
			b.file, err = os.CreateTemp("", "gitea-upload-buffer-")
```

It would surprise users if they see a lot of "/tmp/buffer-123456"

wxiaoguang replied:
And could we reuse `Packages.ChunkedUploadPath` as the tmp dir? Maybe rename `ChunkedUploadPath` to `UploadTmpPath`, etc.

KN4CK3R replied:
This module is not necessarily bundled with packages so I have chosen the os temp path. I could use `setting.Repository.Upload.TempPath` as directory but that does not feel correct too.

wxiaoguang replied:
Yup, maybe we can add the prefix for the temp file, and then unify the tmp path in gitea in future.

wxiaoguang commented on modules/notification/webhook/webhook.go:
```diff
@@ -855,3 +856,33 @@ func (m *webhookNotifier) NotifySyncCreateRef(pusher *user_model.User, repo *rep
+
+func notifyPackage(sender *user_model.User, pd *packages_model.PackageDescriptor, action api.HookPackageAction) {
+	if pd.Repository == nil {
+		// TODO
```
Add a warning log here?

KN4CK3R replied:
I will add a link to #17940.

wxiaoguang commented on models/packages/package_property.go:
```diff
@@ -0,0 +1,70 @@
+
+// PackageProperty represents a property of a package version or file
+type PackageProperty struct {
+	ID      int64        `xorm:"pk autoincr"`
+	RefType PropertyType `xorm:"INDEX NOT NULL"`
+	RefID   int64        `xorm:"INDEX NOT NULL"`
+	Name    string       `xorm:"INDEX NOT NULL"`
```
Not sure whether it is correct that if a package has duplicated property names

wxiaoguang commented:
That's almost the final review from my side. As long as I haven't looked into the package/meta details of each package system, I think it's fine to get it merged after these nits get resolved then eat our own dog food.

KN4CK3R commented on models/packages/package_file.go:
```diff
@@ -0,0 +1,195 @@
+
+	if opts.VersionID != 0 {
+		cond = cond.And(builder.Eq{"package_file.version_id": opts.VersionID})
+	} else if opts.OwnerID != 0 && opts.PackageType != "" && opts.PackageType != "all" {
```
Currently this method is used with `VersionID != 0` or `OwnerID != 0 && PackageType != ""` so the `!= "all"` is possibly not needed (copied from the version search). What do you think about this change which will it make more flexible:
```diff
        if opts.VersionID != 0 {
                cond = cond.And(builder.Eq{"package_file.version_id": opts.VersionID})
-       } else if opts.OwnerID != 0 && opts.PackageType != "" && opts.PackageType != "all" {
+       } else if opts.OwnerID != 0 || (opts.PackageType != "" && opts.PackageType != "all") {
+               var versionCond builder.Cond = builder.Eq{
+                       "package_version.is_internal": false,
+               }
+               if opts.OwnerID != 0 {
+                       versionCond = versionCond.And(builder.Eq{"package.owner_id": opts.OwnerID})
+               }
+               if opts.PackageType != "" && opts.PackageType != "all" {
+                       versionCond = versionCond.And(builder.Eq{"package.type": opts.PackageType})
+               }
+
                in := builder.
                        Select("package_version.id").
                        From("package_version").
                        Join("INNER", "package", "package.id = package_version.package_id").
-                       Where(builder.Eq{
-                               "package.owner_id":            opts.OwnerID,
-                               "package.type":                opts.PackageType,
-                               "package_version.is_internal": false,
-                       })
+                       Where(versionCond)
 
                cond = cond.And(builder.In("package_file.version_id", in))
        }
```

KN4CK3R commented:
Thank you for the review @wxiaoguang 👍

wxiaoguang reviewed (approved):
No blocker from my side. I still hope we can have a clear prefix for temp files though. 😊

wxiaoguang commented:
CI can pass (failure is unrelated)

@go-gitea/maintainers Does anyone have more reviews? If no more reviews, maybe we can merge in one or two days.

delvh reviewed (commented):
Holy, I think I've never been sitting so long on a single review as I have now.
About four hours, and I reviewed only the easy 60%...
But before I forget the rest or this PR gets merged, here are a few points I noticed.

KN4CK3R commented on docs/content/doc/packages/generic.en-us.md:
````diff
@@ -0,0 +1,80 @@
+To download a generic package perform a HTTP GET operation.
+
+```
+GET https://gitea.example.com/api/packages/{owner}/generic/{package_name}/{package_version}/{file_name}
````
In the current form a generic package is only a single file.

KN4CK3R commented on docs/content/doc/packages/maven.en-us.md:
````diff
@@ -0,0 +1,110 @@
+Afterwards run:
+
+```shell
+mvn install
````
I don't know how the Maven package manager works but both GitHub and GitLab use that in their documentation.

https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-apache-maven-registry#installing-a-package
https://docs.gitlab.com/ee/user/packages/maven_repository/#install-a-package

khmarbaise replied:
Downloading of dependencies is done automatically without supplemental command. If you start a build via `mvn package` everything will be downloaded what is needed. 

An `mvn install` means to install the created artifact within your local cache `$HOME/.m2/repository` which means some other project on the same machine can consume those artifacts.
If you like to distribute artifacts to a wider audience it is needed to deploy them via `mvn deploy` which is defined by `<distributionManagement>..</distributionManagement>`...this contains usually a definition of a repository manager...

KN4CK3R commented on templates/package/shared/versionlist.tmpl:
```diff
@@ -0,0 +1,32 @@
+	<p><a href="{{.PackageDescriptor.PackageWebLink}}">{{.PackageDescriptor.Package.Name}}</a> / <strong>{{.i18n.Tr "packages.versions"}}</strong></p>
+	<form class="ui form ignore-dirty">
+		<div class="ui fluid action input">
+			<input name="q" value="{{.Query}}" placeholder="{{.i18n.Tr "explore.search"}}..." autofocus>
```
I understand but that's how every search form in Gitea currently works. Search for `"explore.search"` and they all use the three dots. Maybe we can change it in an other PR.

KN4CK3R commented on templates/admin/packages/list.tmpl:
```diff
@@ -0,0 +1,97 @@
+		{{.i18n.Tr "packages.settings.delete"}}
+	</div>
+	<div class="content">
+		{{.i18n.Tr "packages.settings.delete.notice" `<span class="name"></span>` `<span class="dataVersion"></span>` | Safe}}
```
That's a template and they get filled afterwards:
`<td><a class="delete-button" href="" data-url="{{$.Link}}/delete?page={{$.Page.Paginater.Current}}&sort={{$.SortType}}" data-id="{{.Version.ID}}" data-name="{{.Package.Name}}" data-data-version="{{.Version.Version}}">{{svg "octicon-trash"}}</a></td>`

From `data-name` and `data-data-version` (yes, data-data).

KN4CK3R commented on modules/structs/package.go:
```diff
@@ -0,0 +1,33 @@
+)
+
+// Package represents a package
+type Package struct {
```
Those are API structs not database structs.

KN4CK3R commented on routers/api/packages/helper/helper.go:
```diff
@@ -0,0 +1,38 @@
+	if status == http.StatusInternalServerError {
+		log.ErrorWithSkip(1, message)
+
+		if setting.IsProd && !(ctx.Doer != nil && ctx.Doer.IsAdmin) {
```
Same code is here:
https://github.com/go-gitea/gitea/blob/97625b44e7ec96a99b78e910ddcfdabb687f68b3/modules/context/api.go#L103

Should be changed too.

KN4CK3R commented on services/cron/tasks_basic.go:
```diff
@@ -139,6 +140,20 @@ func registerCleanupHookTaskTable() {
+	RegisterTaskFatal("cleanup_packages", &OlderThanConfig{
+		BaseConfig: BaseConfig{
+			Enabled:    true,
+			RunAtStart: true,
```
Are you talking about #19221? If yes, that's already changed.

delvh replied:
Ah, that value is simply the default value if the setting is missing, correct?
In that case ignore my comment, I thought it would mean that it would always run at start.

KN4CK3R commented on options/locale/locale_en-US.ini:
```diff
@@ -488,7 +488,9 @@ auth_failed = Authentication failed: %v
 
 still_own_repo = "Your account owns one or more repositories; delete or transfer them first."
 still_has_org = "Your account is a member of one or more organizations; leave them first."
+still_own_packages = Your account owns one or more packages; delete them first.
```
I used the same style as the lines above. Should I still change it?

KN4CK3R commented on docs/content/doc/packages/maven.en-us.md:
```diff
@@ -0,0 +1,110 @@
+    <url>https://gitea.example.com/api/packages/{owner}/maven</url>
+  </repository>
+</repositories>
+<distributionManagement>
+  <repository>
+    <id>gitea</id>
+    <url>https://gitea.example.com/api/packages/{owner}/maven</url>
+  </repository>
+  <snapshotRepository>
+    <id>gitea</id>
+    <url>https://gitea.example.com/api/packages/{owner}/maven</url>
+  </snapshotRepository>
+</distributionManagement>
```
GitLab tells you to add this: https://gitlab.com/sabrams/maven-practice-test/-/packages/5171113
Is it needed for the snapshot download?

delvh replied:
As far as I know it isn't.

wxiaoguang replied:
Yup, it's not needed for downloading. 

While I think it's fine to keep it here. A maven user should know what it means and they can choose what they need to write into their local config files.

tboerger commented on docs/content/doc/packages/composer.en-us.md:
````diff
@@ -0,0 +1,120 @@
+     https://gitea.example.com/api/packages/testuser/composer?version=1.0.3
+```
+
+The server reponds with the following HTTP Status codes.
````
responds

tboerger reviewed (commented):
Minimal docs typos

tboerger commented:
Beside the implemented providers it could also make sense to implement the Terraform registry api, Terraform module api and the Terraform remote state api.

lunny commented:
> Beside the implemented providers it could also make sense to implement the Terraform registry api, Terraform module api and the Terraform remote state api.

Good idea but let's merge it at first and send other PRs.

wxiaoguang commented:
Last call for reviews. If no code change in one day, let's merge.

wxiaoguang commented on docs/content/doc/packages/nuget.en-us.md:
````diff
@@ -0,0 +1,116 @@
+
+For example:
+
+```shell
````
````suggestion
```
````

tboerger commented:
@wxiaoguang hopefully my comments will be fixed before the merge ;)

wxiaoguang commented:
> @wxiaoguang hopefully my comments will be fixed before the merge ;)

Yup, these document fixes have been committed. Anything more to change?

tboerger commented:
Not from my side, looks good. Even if I'm still against bloating the gitea core more and more with features, but this won't change without a plugin system.

wxiaoguang commented:
Agree to be more careful about minority features that bloating the core code. Gitea should only introduce features which benefit most users.

Personally speaking, I am not optimistic about Gitea can have a plugin system in near future. Many simple and refactoring PRs just get delayed for long time, even months, I can not imagine how the plugin system could be done in a predictable future.

lunny commented:
Let's merge then send patches. It took so long for this PR.

KN4CK3R commented:
A little bit rushed at the end but I'm happy it got merged after that long time working on it.

KN4CK3R commented on docs/content/doc/packages/nuget.en-us.md:
````diff
@@ -0,0 +1,116 @@
+To do so, register the NuGet package registry as symbol source:
+
+```
+https://gitea.example.com/api/packages/{owner}/nuget/symbols
````
You only need this URL to register the registry as a symbol source for example in Visual Studio.

6543 commented:
locked as new issues should be created linked to this ... if something pops up
